sound/soc/sof/sof-audio.c

```diff
@@ -11,6 +11,561 @@
+			route_found = true;
+			break;
+		}
+
+	if (!route_found) {
+		dev_err(sdev->dev, "error: cannot find SOF route for source %s -> %s sink\n",
+			wsource->name, wsink->name);
+		return -EINVAL;
+	}
+
+	/* nothing to do if route is already set up */
+	if (sroute->setup)
+		return 0;
+
+	connect = sroute->private;
+
+	/* send ipc */
+	ret = sof_ipc_tx_message(sdev->ipc, connect->hdr.cmd, connect, sizeof(*connect),
+				 &reply, sizeof(reply));
+	if (ret < 0) {
+		dev_err(sdev->dev, "error: failed to load route source %s -> %s sink\n",
+			wsource->name, wsink->name);
+	} else {
+		dev_dbg(sdev->dev, "route %s -> %s setup complete\n", wsource->name, wsink->name);
+		sroute->setup = true;
+	}
+
+	return ret;
+}
+
+static bool is_sof_widget(struct snd_soc_dapm_widget *widget)
+{
+	switch (widget->id) {
+	case snd_soc_dapm_dai_in:
+	case snd_soc_dapm_dai_out:
+	case snd_soc_dapm_mixer:
+	case snd_soc_dapm_pga:
+	case snd_soc_dapm_buffer:
+	case snd_soc_dapm_scheduler:
+	case snd_soc_dapm_aif_out:
+	case snd_soc_dapm_aif_in:
+	case snd_soc_dapm_src:
+	case snd_soc_dapm_asrc:
+	case snd_soc_dapm_siggen:
+	case snd_soc_dapm_effect:
+	case snd_soc_dapm_mux:
+	case snd_soc_dapm_demux:
+		if (widget->dobj.private)
+			return true;
+
+		return false;
+	default:
+		return false;
+	}
+}
+
+static int sof_setup_pipeline_connections(struct snd_sof_dev *sdev,
+					  struct snd_soc_dapm_widget_list *list, int dir)
+{
+	struct snd_soc_dapm_widget *widget;
+	struct snd_soc_dapm_path *p;
+	int ret;
+	int i;
+
+	/*
+	 * Set up connections between widgets in the sink/source paths based on direction.
+	 * Some non-SOF widgets exist in topology either for compatibility or for the
+	 * purpose of connecting a pipeline from a host to a DAI in order to receive the DAPM
+	 * events. But they are not handled by the firmware. So ignore them.
+	 */
+	if (dir == SNDRV_PCM_STREAM_PLAYBACK) {
+		for_each_dapm_widgets(list, i, widget) {
+			if (!is_sof_widget(widget))
+				continue;
+
+			snd_soc_dapm_widget_for_each_sink_path(widget, p)
+				if (is_sof_widget(p->sink)) {
+					ret = sof_route_setup(sdev, widget, p->sink);
+					if (ret < 0)
+						return ret;
+				}
+		}
+	} else {
+		for_each_dapm_widgets(list, i, widget) {
+			if (!is_sof_widget(widget))
+				continue;
+
+			snd_soc_dapm_widget_for_each_source_path(widget, p)
+				if (is_sof_widget(p->source)) {
+					ret = sof_route_setup(sdev, p->source, widget);
+					if (ret < 0)
+						return ret;
+				}
+		}
+	}
+
+	return 0;
+}
+
+static int sof_widget_list_dai_config_setup(struct snd_sof_dev *sdev,
+					    struct snd_soc_dapm_widget_list *list)
+{
+	struct snd_soc_dapm_widget *widget;
+	struct snd_sof_widget *swidget;
+	struct snd_sof_dai *dai;
+	int ret, i;
+
+	for_each_dapm_widgets(list, i, widget) {
+		swidget = widget->dobj.private;
+		if (!swidget)
+			continue;
+
+		if  (swidget->id == snd_soc_dapm_dai_in || swidget->id == snd_soc_dapm_dai_out) {
```

Comment:
another option like the following one to reduce tab
````c
if  (swidget->id != snd_soc_dapm_dai_in && swidget->id != snd_soc_dapm_dai_out)
    continue;
````

Reply:
A convenience macro like
```c
#define WIDGET_IS_DAI(w) (w == dai_in || w == dai_ou)
```
would probably be quite useful here as this maybe a common check when we manage pipelines.